### PR TITLE
Surfaces days remaining during grace period

### PR DESCRIPTION
**Pull request by gowthamidommety-okta**
When in grace period, the back-end will return an end date in the policy. This should be used to show the number of days remaining for a end user to enroll in a required factor.

Screen Shot 2019-07-26 at 12 08 47 PM

Screen Shot 2019-07-26 at 12 09 32 PM

### CODE REVIEW

File: src/models/AppState.js [hunk 1]
```diff
@@ -93,6 +93,12 @@ function (Okta, Q, Factor, BrowserFeatures, Errors) {
     return Okta.loc('minutes', 'login', [factorLifetimeInMinutes]);
   }
 
+  function getGracePeriodRemainingDays (gracePeriodEndDate) {
+    let endDate = new Date(gracePeriodEndDate).getTime();
```
Comment (magizh-okta) [inline src/models/AppState.js]: nit: both can be consts we dont reuse them.

File: src/EnrollChoicesController.js [hunk 2]
```diff
@@ -47,13 +47,26 @@ function (Okta, FormController, Enums, RouterUtil, FactorList,
         case Enums.ALL_OPTIONAL_SOME_ENROLLED:
         case Enums.HAS_REQUIRED_ALL_REQUIRED_ENROLLED:
           return Okta.loc('enroll.choices.optional', 'login');
+        case Enums.HAS_REQUIRED_SOME_REQUIRED_ENROLLED:
+          const remainingDays = this.options.appState.get('gracePeriodRemainingDays');
+          return (remainingDays >= 0) ? this.getGracePeriodSubtitle(remainingDays) : this.getDefaultSubtitle();
```
Comment (magizh-okta) [inline src/EnrollChoicesController.js]: if gracePeriodRemainingDays is set to null when it is not defined. 

null >= 0   gives => true 

We need to update the logic here.

Comment (gowthamidommety-okta) [reply src/EnrollChoicesController.js]: Yup, changed this. Will push soon

File: src/models/AppState.js (hunk 1 above)
Comment (gowthamidommety-okta) [reply src/models/AppState.js]: Changing this to var to be consistent with the rest of the file

File: src/EnrollChoicesController.js [hunk 3]
```diff
@@ -47,13 +47,27 @@ function (Okta, FormController, Enums, RouterUtil, FactorList,
         case Enums.ALL_OPTIONAL_SOME_ENROLLED:
         case Enums.HAS_REQUIRED_ALL_REQUIRED_ENROLLED:
           return Okta.loc('enroll.choices.optional', 'login');
+        case Enums.HAS_REQUIRED_SOME_REQUIRED_ENROLLED:
+          var remainingDays = this.options.appState.get('gracePeriodRemainingDays');
+          return (Number.isInteger(remainingDays) && remainingDays >= 0) ?
```
Comment (mauriciocastillosilva-okta) [inline src/EnrollChoicesController.js]: I think `gracePeriodRemainingDays` should return an integer all the time, with value 0 or above. At the end, `0` means required right now... Unless we need to know that the value is not actually set, which is where returning `null` will help. But I don't think we need that

File: test/unit/spec/EnrollChoices_spec.js [hunk 4]
```diff
@@ -522,6 +545,97 @@ function (Okta, OktaAuth, Util, EnrollChoicesForm, Beacon, Expect, FactorUtil, R
         });
       });
 
+      Expect.describe('Grace period', function () {
+        beforeEach(function () {
+          var today = new Date('2019-06-25T00:00:00.000Z');
+          jasmine.clock().mockDate(today);
+        });
+        Expect.describe('all factors are required', function () {
```
Comment (mauriciocastillosilva-okta) [inline test/unit/spec/EnrollChoices_spec.js]: nit: `all factors are required and none enrolled`

File: src/EnrollChoicesController.js (hunk 3 above)
Comment (mauriciocastillosilva-okta) [reply src/EnrollChoicesController.js]: talked offline, makes sense as is. `null` means no grace period, and then any number will be handled in `getGracePeriodSubtitle`, where `0` means we still are in grace period

File: test/unit/spec/EnrollChoices_spec.js [hunk 5]
```diff
@@ -522,6 +545,97 @@ function (Okta, OktaAuth, Util, EnrollChoicesForm, Beacon, Expect, FactorUtil, R
         });
       });
 
+      Expect.describe('Grace period', function () {
+        beforeEach(function () {
```
Comment (magizh-okta) [inline test/unit/spec/EnrollChoices_spec.js]: should we install/uninstall  clock so we just restrict the jasmine clock behavior to this block  ?

Comment (gowthamidommety-okta) [reply test/unit/spec/EnrollChoices_spec.js]: Filed OKTA-241179 to look into this

**Pull request merged**